### ISSUE TITLE
Extreme fit time when using a subset of the features

### ISSUE THREAD

**Issue by LHBO** [PriorLabs/TabPFN#731]
### Describe the bug

TabPFN (both classifier and regressor) is very slow when training on few features, often categorical or containing only integers. I have upgraded to the latest version of TabPFN and I am running on a NVIDIA RTX 4000 SFF.

In the sense that fitting on the full dataset (all instances and features) takes 3 seconds, but only fitting on a subset of the features can take minutes and hours. The latter means that using TabPFN is not feasible. I am wondering if I am doing something wrong or if there are some reasons for the long running time.

### Steps/Code to Reproduce

```
from sklearn.datasets import load_breast_cancer
from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.model_selection import train_test_split

from tabpfn import TabPFNClassifier
from tabpfn.constants import ModelVersion

from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
adult = fetch_ucirepo(id=2) 
  
# data (as pandas dataframes) 
X = adult.data.features 
y = adult.data.targets 
  
# Split the dataset into training and testing sets
X_train_adult, X_test_adult, y_train_adult, y_test_adult = train_test_split(X, y, test_size=0.1, random_state=42)
y_train_adult = y_train_adult.to_numpy().ravel()
y_test_adult = y_test_adult.to_numpy().ravel()
X_train_adult = X_train_adult.to_numpy()
X_test_adult = X_test_adult.to_numpy()

# Initialize a classifier
clf = TabPFNClassifier(device="cuda") 
clf.fit(X_train_adult, y_train_adult) # Takes 3 seconds

# Predict probabilities
prediction_probabilities = clf.predict_proba(X_test_adult)
print("ROC AUC:", roc_auc_score(y_test_adult, prediction_probabilities[:, 1]))

# Predict labels
predictions = clf.predict(X_test_adult)
print("Accuracy", accuracy_score(y_test_adult, predictions))

####################################
### HERE IS THE PROBLEMATIC PART ###
####################################
# Only use a subset of the features
X_train_adult_one_feature = X_train_adult[:, 11].reshape(-1, 1)
X_test_adult_one_feature = X_test_adult[:, 11].reshape(-1, 1)

# Retrain the classifier with only one feature (This takes many hours)
use_n_instances = 1000
clf.fit(X_train_adult_one_feature[0:use_n_instances], y_train_adult[0:use_n_instances])

# 100 values takes 0.8 seconds
# 500 values takes 8 seconds
# 1000 values takes 33 seconds
# 2000 values takes 2 minutes and 10 seconds
# The time seems to grow exponentially with the number of instances used for training when only having one feature.
# or several features that are categorical/integer.


####################################
# Take two features
X_train_adult_two_features = X_train_adult[:, [0, 7]]
X_test_adult_two_features = X_test_adult[:, [0, 7]]

# Retrain the classifier with two features (which is faster but still much slower than using all features and all instances)
use_n_instances = 10000
clf.fit(X_train_adult_two_features[0:use_n_instances], y_train_adult[0:use_n_instances]) 

# 100 values takes 0.7 seconds
# 500 values takes 0.5 seconds
# 1000 values takes 0.9 seconds
# 2000 values takes 2.6 seconds
# 5000 values takes 14.8 seconds
# 10000 values takes 59 seconds
```

### Expected Results

I would have expected the running time to be about the same or shorter as I am using less features.

### Actual Results

The running time increases exponentially when using more of the training instances in fit.

### Versions

```shell
Collecting system and dependency information...
PyTorch version: 2.9.1+cu128
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jan  8 2026, 06:52:19) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-157-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA RTX 4000 SFF Ada Generation
GPU 1: NVIDIA RTX 4000 SFF Ada Generation
GPU 2: NVIDIA TITAN X (Pascal)
GPU 3: NVIDIA TITAN X (Pascal)

Nvidia driver version: 580.95.05
cuDNN version: Probably one of the following:
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.8.1.1
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.1.1
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.1.1
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.1.1
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.1.1
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.1.1
/lokal_uten_backup/cuda/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.1.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.6/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.3.2
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn.so.8.8.0
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.8.0
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.8.0
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.8.0
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.8.0
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.8.0
/lokal_uten_backup/cuda/cuda-11.7/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn.so.4.0.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.5.0.5
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn.so.8.8.0
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.8.0
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.8.0
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.8.0
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.8.0
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.8.0
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
CPU family:                              6
Model:                                   63
Thread(s) per core:                      2
Core(s) per socket:                      8
Socket(s):                               2
Stepping:                                2
CPU max MHz:                             2400.0000
CPU min MHz:                             1200.0000
BogoMIPS:                                4800.37
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm arat pln pts md_clear flush_l1d
Virtualization:                          VT-x
L1d cache:                               512 KiB (16 instances)
L1i cache:                               512 KiB (16 instances)
L2 cache:                                4 MiB (16 instances)
L3 cache:                                40 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-7,16-23
NUMA node1 CPU(s):                       8-15,24-31
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                       Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:                  Mitigation; PTI
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected

Dependency Versions:
--------------------
tabpfn: 6.3.0
torch: 2.9.1
numpy: 1.26.4
scipy: 1.15.3
pandas: 2.3.3
scikit-learn: 1.6.1
typing_extensions: 4.15.0
einops: 0.8.1
huggingface-hub: 1.1.5
```

**Comment by alanprior** [comment]
Hello @LHBO and thank you for finding a bug!

The team was able to replicate it and identify its source, and we'll aim to fix this.

For now, could you try using:
`clf = TabPFNClassifier(device="cuda", inference_config={"FINGERPRINT_FEATURE": False})`

This should hopefully unblock you.

**Comment by LHBO** [comment]
Hey, @alanprior. Thank you very much for the rapid reply and I can confirm that this fixes the long running time. 
I want to make a note that I see a drop in accuracy though when using this fix.

**Comment by alanprior** [comment]
Thanks a lot @LHBO. This is insightful, and it makes sense. Fortunately this should be fixable :)

**Comment by noahho** [comment]
Proposed Solution: Refactor the collision handling in `add_fingerprint_features_step.py`. Instead of the current while loop, implement a counter-based approach (similar to logic found in AutoGluon) to handle unique identifiers/hashes more efficiently.

Priority: Medium/High. This affects a common use case (feature selection/subsets) and datasets with specific characteristics (binary/integer features).

**Comment by noahho** [comment]
This seems reasonable, we are making sure the hash resolution stays deterministic. Here is the minimal fix. The primary change replaces the $O(N^2)$ linear probing loop with a dictionary-based counter (hash_counter) to track occurrences of identical rows. This reduces the complexity to $O(N)$ for datasets with many duplicates or feature subsets.
```
"""Add Fingerprint Features Step."""

from __future__ import annotations

import hashlib
from collections import defaultdict  # Added import
from typing_extensions import override

import numpy as np
import torch

from tabpfn.preprocessing.pipeline_interfaces import (
    FeaturePreprocessingTransformerStep,
)
from tabpfn.utils import infer_random_state

_CONSTANT = 10**12


def _float_hash_arr(arr: np.ndarray) -> float:
    _hash = int(hashlib.sha256(arr.tobytes()).hexdigest(), 16)
    return _hash % _CONSTANT / _CONSTANT


class AddFingerprintFeaturesStep(FeaturePreprocessingTransformerStep):
    """Adds a fingerprint feature to the features based on hash of each row.

    If `is_test = True`, it keeps the first hash even if there are collisions.
    If `is_test = False`, it handles hash collisions by counting up and rehashing
    until a unique hash is found.
    The idea is basically to add a random feature to help the model distinguish between
    identical rows. We use hashing to make sure the result does not depend on the order
    of the rows.
    """

    def __init__(self, random_state: int | np.random.Generator | None = None):
        super().__init__()
        self.random_state = random_state

    @override
    def _fit(
        self, X: np.ndarray | torch.Tensor, categorical_features: list[int]
    ) -> list[int]:
        _, rng = infer_random_state(self.random_state)
        self.rnd_salt_ = int(rng.integers(0, 2**16))
        return [*categorical_features]

    @override
    def _transform(  # type: ignore
        self,
        X: np.ndarray | torch.Tensor,
        *,
        is_test: bool = False,
    ) -> np.ndarray | torch.Tensor:
        X_det = X.detach().cpu().numpy() if isinstance(X, torch.Tensor) else X

        # no detach necessary for numpy
        X_h = np.zeros(X.shape[0], dtype=X_det.dtype)
        if is_test:
            # Keep the first hash even if there are collisions
            salted_X = X_det + self.rnd_salt_
            for i, row in enumerate(salted_X):
                h = _float_hash_arr(row + self.rnd_salt_)
                X_h[i] = h
        else:
            # Handle hash collisions by counting up and rehashing
            seen_hashes = set()
            # Map initial hash -> next candidate offset to avoid O(N^2) on duplicates
            hash_counter = defaultdict(int)
            
            salted_X = X_det + self.rnd_salt_
            for i, row in enumerate(salted_X):
                # Calculate the base hash to identify the row content
                h_base = _float_hash_arr(row)
                
                # Start checking from the last known count for this row content
                add_to_hash = hash_counter[h_base]
                
                h = _float_hash_arr(row + add_to_hash)
                
                # Resolve remaining collisions (if row+k accidentally collides with another row)
                while h in seen_hashes and not np.isnan(row).all():
                    add_to_hash += 1
                    h = _float_hash_arr(row + add_to_hash)
                
                X_h[i] = h
                seen_hashes.add(h)
                
                # Update counter so the next identical row starts checking from the new offset
                hash_counter[h_base] = add_to_hash + 1

        if isinstance(X, torch.Tensor):
            return torch.cat(
                [X, torch.from_numpy(X_h).float().reshape(-1, 1).to(X.device)], dim=1
            )
        else:  # noqa: RET505
            return np.concatenate([X, X_h.reshape(-1, 1)], axis=1)


__all__ = [
    "AddFingerprintFeaturesStep",
]
```

**Comment by alanprior** [comment]
Proposed PR following @noahho's suggestion:
https://github.com/PriorLabs/TabPFN/pull/742

**Comment by noahho** [comment]
Fixed with #742
